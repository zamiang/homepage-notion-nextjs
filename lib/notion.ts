import { Client } from '@notionhq/client';

const notion = new Client({
  auth: process.env.NOTION_TOKEN,
});

export const getDatabase = async (databaseId: string) => {
  const response = await notion.databases.query({
    database_id: databaseId,
  });
  return response.results;
};

export const getPage = async (pageId: string) => {
  const response = await notion.pages.retrieve({ page_id: pageId });
  return response;
};

export const getPageBySlug = async (slug: string, databaseId: string) => {
  const response = await notion.databases.query({
    database_id: databaseId,
    page_size: 1,
    filter: {
      property: 'Slug',
      text: {
        equals: slug,
      },
    },
    sorts: [{ timestamp: 'last_edited_time', direction: 'ascending' }],
  });
  return response.results[0];
};

export const getBlocks = async (blockId: string) => {
  const response = await notion.blocks.children.list({
    block_id: blockId,
    page_size: 150,
  });
  return response.results;
};
