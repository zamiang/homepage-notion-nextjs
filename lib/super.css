html {
  -webkit-text-size-adjust: 100%; /* Prevent font scaling in landscape while allowing user zoom */
}

/* --- Change default colors --- */
:root {
  --color-bg-default: #f9f6f0 !important;
  --color-text-default: #4b5673 !important;
  --color-text-default-light: #4b5673 !important;
  --color-text-gray: #959ba9 !important;
  --collection-card-cover-size-large: 210px !important;
}
@media screen and (min-width: 50em) {
  :root {
    --collection-card-cover-size-large: 270px !important;
  }
}

/* --- TYPOGRAPHY --- */

body {
  font-family: reason-new, "Work Sans", sans-serif !important;

  /* Make the text nice and crisp */
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  /* Improve kerning for good measure */
  text-rendering: optimizeLegibility;
}

strong {
  font-weight: 500 !important;
}

/* Styling for page names */
.notion-header__title {
  font-weight: 500 !important;
  font-size: 0.75rem !important;
  line-height: 1.25 !important;
  min-height: auto !important;
  text-transform: uppercase !important;
  letter-spacing: 0.1em !important;
  padding: 0 !important;
  display: none;
}
.notion-header__content > div {
  margin-bottom: 0px !important;
}
.notion-header {
  margin-bottom: 0px !important;
  padding-top: 32px !important;
}
@media screen and (min-width: 680px) {
  .notion-header {
    padding-top: 80px !important;
  }
}

/* Adjust line-height and letter-spacing for body copy */
.notion-semantic-string {
  line-height: 1.7 !important;
  letter-spacing: -0.01em !important;
  font-size: 18px;
}

/* --- Heading styles --- */

/* H1 */
h1.notion-heading,
.notion-page__properties .notion-property__text {
  font-family: psfournier-std, "Nanum Myeongjo", serif !important;
  font-weight: 200 !important;
  font-size: 1.875rem !important;
}

h1.notion-heading .notion-semantic-string,
.notion-page__property .notion-property__text .notion-semantic-string {
  font-size: 1.3rem !important;
  line-height: 1.27 !important;
}

.notion-page__property .notion-property__text .notion-semantic-string {
  margin: 10px auto;
  padding: 0 10%;
  text-align: center;
}


h1.notion-heading .notion-semantic-string,
.notion-page__property:first-child .notion-property__text .notion-semantic-string {
  line-height: 1.27 !important;
  letter-spacing: -0.01em !important;
  font-size: 1.875rem !important;
  margin: 0 auto 15px auto;
}
@media screen and (min-width: 50em) {
  h1.notion-heading .notion-semantic-string,
.notion-page__property:first-child .notion-property__text .notion-semantic-string {
    font-size: 2.55rem !important;
  }
}

/* H2 */
h2.notion-heading {
  font-family: psfournier-std, "Nanum Myeongjo", serif !important;
  font-weight: 200 !important;
  letter-spacing: -0.01em !important;
}

/* H3 */
h3.notion-heading {
  font-weight: 500 !important;
  margin-bottom: 0.25em !important;
}

/* Remove padding */
.notion-heading {
  padding: 0 !important;
}

/* Adjust line height for H2 and H3 */
.notion-heading .notion-semantic-string {
  line-height: 1.3 !important;
}

/* Ensure opacity remains at 1 when a heading is a link */
.notion-heading .notion-semantic-string .link {
  opacity: 1 !important;
}

.notion-text__content {
  padding: 8px 0 !important;
}

/* --- CONTENT STYLING --- */

/* --- Remove bottom page padding and adjust page padding for smaller screens --- */
.super-content {
  padding-bottom: 40px !important;
}
@media screen and (min-width: 415px) and (max-width: 860px) {
  .super-content,
  .notion-header__content {
    padding-left: 46px !important;
    padding-right: 46px !important;
  }
}

/* --- Remove additional spacing on mobile that is created by column layout on desktop --- */
@media screen and (max-width: 680px) {
  .notion-column {
    padding: 0px !important;
  }
}

/* --- Control heading spacing in column layout --- */
.notion-column > .notion-heading {
  margin-top: 7px !important;
}

.notion-header__icon-wrapper {
  padding: 2px !important;
  bottom: 8px !important;
}

/* --- Adjust quote styling so it can be used as a standfirst --- */
.notion-quote {
  border-left: 0px !important;
  padding: 0 !important;
  margin: 0.5rem 0 1rem !important;
}

/* --- Change handling of full width images --- */
.notion-image.page-width > div {
  display: block !important;
  margin-bottom: 6px !important;
}

.notion-image.full-width img {
  object-fit: fill !important;
  height: auto !important;
}

/* --- Change Image caption alignment --- */
.notion-image.page-width > figcaption,
.notion-image.full-width > figcaption {
  text-align: center !important;
}

/* --- Photos LISTING --- */

/* Remove header of the gallery table */
.notion-collection__header {
  display: none !important;
}

/* Stop card from breaking out of layout on screens smaller than 370px */
.notion-collection-gallery {
  display: block !important;
  border-top: none !important;
}

@media screen and (min-width: 370px) {
  /* Adjust spacing of gallery grid */
  .notion-collection-gallery {
    display: grid !important;
    grid-gap: 46px !important;
    gap: 46px !important;
    grid-row-gap: 66px !important;
  }
}

/* Override default card styling */
.notion-collection-card {
  border-radius: 0 !important;
  box-shadow: none !important;
  margin-bottom: 44px !important;
  background: transparent !important;
}
@media screen and (min-width: 370px) {
  .notion-collection-card {
    margin-bottom: 0 !important;
  }
}

.notion-collection-card:hover {
  background: transparent !important;
}
.notion-collection-card__cover.large,
.notion-collection-card__cover.large > div,
.notion-collection-card__cover.large img,
.notion-collection-card__cover.medium,
.notion-collection-card__cover.medium > div,
.notion-collection-card__cover.medium img {
  height: auto !important;
  max-height: none !important;
}
.notion-collection-card__cover.large > div {
  position: relative !important;
}
.notion-collection-card__cover {
  height: auto !important;
  min-height: none !important;
  border-bottom: none !important;
}
.notion-collection-card__cover img {
  position: relative !important;
  height: auto !important;
  width: 100% !important;
  border-radius: 0 !important;
  object-fit: contain !important;
  object-position: initial !important;
  padding-bottom: 0 !important;
}
.notion-collection-card__content {
  padding-bottom: 0 !important;
  padding-left: 0 !important;
  font-size: 1rem !important;
  overflow: hidden;
}
.notion-collection-card__content.no-title {
  padding-top: 0 !important;
}

/* Hides page title */
.notion-property__title {
  display: none !important;
}

/* Display text property as a heading */
.notion-collection-card__property .notion-semantic-string {
  display: block !important;
  margin: 0.6em 0 0.4em !important;
  font-family: psfournier-std, "Nanum Myeongjo", serif !important;
  font-weight: 200 !important;
  font-size: 1.5rem !important;
  line-height: 1.3 !important;
  letter-spacing: -0.01em !important;
  white-space: normal !important;
}

/* Add underline to title on hover of project card */
.notion-collection-card.gallery
  .notion-collection-card__property
  .notion-semantic-string
  span {
  border-bottom: 1px solid transparent !important;
  transition: border-color 0.2s ease-in;
}
.notion-collection-card.gallery:hover
  .notion-collection-card__property
  .notion-semantic-string
  span {
  border-bottom: 1px solid var(--color-text-gray) !important;
}

/* Increase spacing between project meta data */
.notion-collection-card__property {
  margin: 3px 0 !important;
}

/* Stop the tags from truncating */
.notion-property__select {
  display: inline !important;
  white-space: normal !important;
  overflow: auto !important;
}

/* Adjust pill styling */
.notion-pill {
  font-size: 1rem !important;
  border-radius: 0px !important;
  line-height: 24px !important;
  margin: 2px 0 !important;
  padding: 0 !important;
  display: inline !important;
  white-space: normal !important;
  text-overflow: initial !important;
}

/* Adds comma before each item */
.notion-pill::before {
  content: ", ";
  white-space: pre;
}

/* Removes comma before first item */
.notion-pill.first::before {
  content: "";
}

/* Make all but the default colour display as a lighter colour */
.pill-default {
  background: none !important;
}
.pill-gray,
.pill-brown,
.pill-orange,
.pill-yellow,
.pill-green,
.pill-blue,
.pill-purple,
.pill-pink,
.pill-red {
  background: none !important;
  color: var(--color-text-gray) !important;
}

/* --- PROJECT PAGE --- */

/* Hide cover image from page headers */
.notion-header__cover {
  display: none !important;
}
.notion-header__content.no-icon.has-cover .notion-header__title-wrapper {
  margin-top: 0 !important;
}
article[id*="photos-"]
  .notion-page__property
  > .property-49775679
  .notion-pill {
  font-weight: 600 !important;
  font-size: 0.75rem !important;
  line-height: 1.25 !important;
  min-height: auto !important;
  text-transform: uppercase !important;
  letter-spacing: 0.1em !important;
  padding: 0 !important;
}

/* Amend the display of Notionâ€™s page properties to act as header and meta data */

/* Adjust spacing between header text elements */
.notion-page__properties {
  margin: 16px 0 0 0;
}
@media screen and (min-width: 680px) {
  .notion-page__properties {
    margin: 30px 0 60px 0;
  }
}

.notion-property__text .notion-semantic-string {
  margin-bottom: 16px;
}

/* Display as single column */
.notion-page__property {
  grid-template-columns: 1fr !important;
}

/* Remove default styling of properties */
.notion-page__properties .notion-property {
  border-radius: 0 !important;
  padding: 0 !important;
  width: 100%;
}
@media screen and (min-width: 50em) {
  .notion-page__properties .notion-property {
    font-size: 1rem !important;
  }
}

/* Hide property name */
.notion-page__property-name {
  display: none !important;
}

/* Hide divider */
.notion-page__properties .notion-divider {
  display: none;
}

/* - - - - - - - - - - - - - - - - - -

CUSTOMISING YOUR BLOG:

All the styles below are appended with a conditional target, it looks like this [id*="writing"]. What that is doing is looking for the presence of the string â€˜writingâ€™ in an id. That id is generated when you give a page a Pretty URL in Super. So long as the listing page and the article pages contain â€˜writingâ€™ in the pretty URL everything below will work fine.

If you prefer to use a different term to writing e.g. blog, thoughts, journal then youâ€™ll need to first use that term for your Pretty URLs then find and replace all instances of "writing" with your chosen term. Hereâ€™s an example:

Listing view: [id*="writing"] becomes [id*="journal"]
Article view: [id*="writing-"] becomes [id*="journal-"]

 - - - - - - - - - - - - - - - - - - */

/* --- Article listing --- */

/* Style for listing of articles */
.notion-collection-list {
  display: block !important;
  padding: 0 !important;
  border-top: 0 !important;
}

/* Center aligns list on article listing page */
[id*="writing"] .notion-collection-list {
  margin: 0 auto;
}

/* Set spacing between articles */
.notion-collection-list__item {
  padding: 0 !important;
  margin: 1em 0 3em 0 !important;
  border-radius: 0 !important;
  display: block !important;
}
@media screen and (min-width: 32em) {
  .notion-collection-list__item {
    margin: 2em 0 3em 0 !important;
  }
  [id*="writing"] .notion-collection-list__item {
    margin-bottom: 5em !important;
  }
}
.notion-collection-list__item:last-child {
  margin-bottom: 2em !important;
}

/* Hover styles for listing of articles */
.notion-collection-list__item:hover {
  background-color: transparent !important;
}
.notion-collection-list__item:hover .property-6f547c5f .notion-semantic-string {
  border-bottom: 1px solid var(--color-text-gray) !important;
}

/* Restyle meta data of each article in the list */
.notion-collection-list__item-property {
  padding: 0 !important;
}
.notion-collection-list__item-property .notion-semantic-string {
  white-space: normal !important;
}
.notion-collection-list__item-content,
[id*="writing"]
  .notion-collection-list__item-property
  .notion-property__select {
  display: block !important;
}
.notion-property__title {
  display: block !important;
}

/* Hides icon */
.notion-property__title .notion-property__title__icon-wrapper {
  display: none !important;
}

/* Hides page title */
.notion-collection-list__item .notion-property__title .notion-semantic-string {
  display: none !important;
}

/* Resets font size of date */
.notion-collection-list__item-property
  .notion-property__text
  .notion-semantic-string {
  font-size: 1rem !important;
  line-height: 1.7 !important;
}

/* Resets font size of date */
.notion-collection-list__item-property .notion-property__date {
  font-size: 1rem !important;
}

.notion-property.notion-property__date .notion-semantic-string {
  text-align: center;
  margin: 0 auto;
}

/* Resets spacing and padding on date */
.notion-collection-list__item-property
  .notion-property__date
  .notion-semantic-string {
  min-height: auto !important;
  padding: 0 !important;
}

/* Set colour and size of date */
.notion-collection-list__item-property
  .notion-property__date
  .notion-semantic-string
  .date {
  color: var(--color-text-gray) !important;
  font-size: 0.875rem !important;
}

/* Styles for title */
.notion-collection-list__item-property .property-6f547c5f {
  margin: 0.5em 0 1em !important;
}
.notion-collection-list__item-property
  .property-6f547c5f
  .notion-semantic-string {
  font-family: psfournier-std, "Nanum Myeongjo", serif !important;
  font-weight: 200 !important;
  font-size: 1.875rem !important;
  letter-spacing: -0.01em !important;
  line-height: 1.25 !important;
  border-bottom: 1px solid transparent !important;
  transition: border-color 0.2s ease-in;
}
@media screen and (min-width: 32em) {
  .notion-collection-list__item-property
    .property-6f547c5f
    .notion-semantic-string {
    font-size: 2.111rem !important;
    line-height: 1.15 !important;
  }
}

/* Uses Notionâ€™s file property to display featured image for article */
.notion-collection-list__item-property .notion-property__file a > div,
.notion-collection-list__item-property .notion-property__file img {
  width: 100% !important;
  height: auto !important;
  max-height: none !important;
  margin-right: 0 !important;
}
.notion-collection-list__item-property .notion-property__file a > div > div,
.notion-collection-list__item-property .notion-property__file img {
  position: relative !important;
}
.notion-collection-list__item-property .notion-property__file img {
  margin-top: 0.5em !important;
  margin-bottom: 1.5em !important;
}

/* --- Article page --- */

/* Set width of body and center it */
#block-writing, #block-photos,
 [id*="writing-"] .notion-page__properties,
[id*="writing-"] .notion-text,
[id*="writing-"] .notion-heading,
[id*="writing-"] .notion-quote,
[id*="writing-"] .notion-divider,
[id*="writing-"] .notion-collection,
[id*="writing-"] .notion-caption,
[id*="writing-"] .notion-bulleted-list,
[id*="writing-"] .notion-numbered-list {
  width: 100%;
  max-width: 730px !important;
  margin-left: auto !important;
  margin-right: auto !important;
}
.notion-collection-table__cell.title, .notion-collection-table__head-cell.title {
  padding-left: 0px;
}
.notion-collection-table__head-cell {
  text-decoration: underline;
}
.notion-collection-table__head-cell.number {
  text-align: right;
}
[id*="writing-"] .notion-column-list {
  width: 100%;
  max-width: 730px !important;
  margin-left: auto !important;
  margin-right: auto !important;
}
[id*="writing-"] .notion-page__properties {
  margin-bottom: 16px !important;
}
@media screen and (min-width: 680px) {
  [id*="writing-"] .notion-page__properties {
    margin-bottom: 40px !important;
  }
}

/* Style date */
[id*="writing-"] .notion-property__date .notion-semantic-string .date {
  color: var(--color-text-gray) !important;
  font-size: 1rem !important;
}
@media screen and (min-width: 32em) {
  [id*="writing-"] .notion-property__date .notion-semantic-string .date {
    font-size: 0.875rem !important;
  }
}

/* Style of page title can be found with heading styles */
/* Style body titles */
[id*="writing-"] h2.notion-heading {
  font-family: reason-new, "Work Sans", sans-serif !important;
  font-weight: 500 !important;
  letter-spacing: 0.015em !important;
  margin-top: 1.25em !important;
  margin-bottom: 0.25em !important;
  text-indent: -0.05em !important;
}
[id*="writing-"] h3.notion-heading {
  margin-top: 1.75rem;
}

/* Style images */
[id*="writing-"] .notion-image {
  margin-bottom: 1.25em !important;
}
@media screen and (min-width: 32em) {
  [id*="writing-"] .notion-image {
    margin-bottom: 2em !important;
  }
}
[id*="writing-"] .notion-image.normal {
  width: 100% !important;
  max-width: 730px !important;
  margin-left: auto !important;
  margin-right: auto !important;
}
[id*="writing-"] .notion-image.normal img {
  margin: 0 auto 0.5em !important;
}

/* Center image caption (affects image captions across the site) */
figcaption {
  text-align: center !important;
  padding: 6px 0 0 !important;
  color: var(--color-text-gray) !important;
}

/* Fix border colour for captions that contain links */
figcaption a {
  opacity: 1 !important;
  border-color: var(--color-text-gray) !important;
}

/* Control space between an image and heading */
[id*="writing-"] .notion-image.page-width + h2.notion-heading {
  margin-top: 1em !important;
}

/* Control space between body text and image */
[id*="writing-"] .notion-text + .notion-image {
  margin-top: 1.25em !important;
}
@media screen and (min-width: 32em) {
  [id*="writing-"] .notion-text + .notion-image {
    margin-top: 2em !important;
  }
}

/* Style quotes */
[id*="writing-"] .notion-quote {
  margin-top: 1.25em !important;
  margin-bottom: 1.25em !important;
  padding-left: 1.5em !important;
  border-left: 4px solid #c7cad2 !important;
}

/* Style lists */
.notion-numbered-list,
.notion-bulleted-list {
  padding-left: 1.25em !important;
  margin-bottom: 1em !important;
}

/* Style dividers to display as centered ellipsis */
[id*="writing-"] .notion-divider {
  border-bottom: none !important;
  margin: 0 auto !important;
  text-align: center;
  font-size: 1.25rem;
}
[id*="writing-"] .notion-divider::before {
  content: "Â· Â· Â·";
}

/* --- MISC --- */

/* Hide notion toggle - a hack to hide child pages, drag them into a toggle and they'll disappear */
.notion-toggle {
  display: none !important;
}

/* --- FOOTER --- */
/* The footer makes use of the standard table view consider whether youâ€™ll need this feature for other aspects of your site before implementing the footer in this way */

/* Add dividing border */
.notion-collection-table__wrapper {
  padding-top: 1em !important;
  margin-top: 3em;
}
/* Reset font size */
.notion-collection-table {
}
/* Hide header row */
.notion-collection-table th {
}
/* Remove borders of cells */
.notion-collection-table td {
}

.notion-collection-table__body .title {
}
/* Reinstate usual styling of links */
.notion-collection-table td .notion-semantic-string .link {
  opacity: 0.7 !important;
  transition: opacity 0.1s ease-in;
}
.notion-collection-table td .notion-semantic-string .link:hover {
  opacity: 1 !important;
}
/* Donâ€™t truncate content on small screens */
.notion-collection-table__cell .notion-semantic-string {
  white-space: normal !important;
}
.super-navbar__logo span {
  font-weight: normal !important;
}

.notion-collection-table__wrapper {
 margin-top: 0px;
}

.notion-collection-table__cell.title>a .notion-semantic-string>span {
  border-bottom: 0px solid;
  pointer-events: none;
}

.notion-collection-table__cell.title>a:hover .notion-semantic-string>span {
  border-bottom: 0px solid;
  pointer-events: none;
}
