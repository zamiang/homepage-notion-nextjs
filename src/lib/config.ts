/**
 * Centralized configuration with environment variable validation
 */

/**
 * Validates and retrieves a required environment variable
 * @throws Error if the environment variable is not set
 */
function requireEnv(key: string): string {
  const value = process.env[key];
  if (!value) {
    throw new Error(`Missing required environment variable: ${key}`);
  }
  return value;
}

/**
 * Retrieves an optional environment variable with a default value
 */
function getEnv(key: string, defaultValue: string): string {
  return process.env[key] || defaultValue;
}

/**
 * Application configuration
 * Validates environment variables on import
 */
export const config = {
  /**
   * Site configuration
   */
  site: {
    url: getEnv('NEXT_PUBLIC_BASE_URL', 'https://www.zamiang.com'),
    title: 'Brennan Moore - Blog',
    description: 'Writing and photos by Brennan Moore',
    author: 'Brennan Moore',
  },

  /**
   * Notion API configuration
   * Note: These are only required when running the cache script or fetching from Notion
   * During static builds, we use cached data, so these can be optional
   */
  notion: {
    token: process.env.NOTION_TOKEN || '',
    dataSourceId: process.env.NOTION_DATA_SOURCE_ID || '',
    photosDataSourceId: process.env.NOTION_PHOTOS_DATA_SOURCE_ID || '',
  },

  /**
   * Cache configuration
   */
  cache: {
    postsFileName: 'posts-cache.json',
    photosFileName: 'photos-cache.json',
  },

  /**
   * Image configuration
   */
  images: {
    minimumCacheTTL: 2678400, // 30 days in seconds
  },
} as const;

/**
 * Validates required Notion environment variables
 * Call this when you need to access Notion API (e.g., in cache script)
 */
export function validateNotionConfig() {
  requireEnv('NOTION_TOKEN');
  requireEnv('NOTION_DATA_SOURCE_ID');
  requireEnv('NOTION_PHOTOS_DATA_SOURCE_ID');
}

/**
 * Type-safe config access
 */
export type Config = typeof config;
