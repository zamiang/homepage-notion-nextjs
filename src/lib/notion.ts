import dotenv from 'dotenv';
dotenv.config();
import { Client } from '@notionhq/client';
import { NotionToMarkdown } from 'notion-to-md';
import { PageObjectResponse } from '@notionhq/client/';
import fs from 'fs';
import path from 'path';
import { ImageBlockObjectResponse } from '@notionhq/client/build/src/api-endpoints';
import { downloadImage, getFilename } from './download-image';

export const notion = new Client({ auth: process.env.NOTION_TOKEN });
export const n2m = new NotionToMarkdown({ notionClient: notion });

n2m.setCustomTransformer('image', async (block) => {
  const { image } = block as ImageBlockObjectResponse;
  const src = image.type === 'external' ? image.external.url : image.file.url;
  const filename = getFilename(src);

  if (true) {
    downloadImage(src);
  }

  return `<figure><img src="/images/${filename}" /></figure>`;
});

export interface Post {
  id: string;
  title: string;
  slug: string;
  coverImage: string;
  date: string;
  excerpt?: string;
  content: string;
  author?: string;
  tags?: string[];
  category?: string;
}

export async function getDatabaseStructure(databaseID: string) {
  const database = await notion.databases.retrieve({
    database_id: databaseID,
  });
  return database;
}

export function getWordCount(content: string): number {
  const cleanText = content
    .replace(/[^\w\s]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
  return cleanText.split(' ').length;
}

export function getPostsFromCache(): Post[] {
  const cachePath = path.join(process.cwd(), 'posts-cache.json');
  if (fs.existsSync(cachePath)) {
    const cache = fs.readFileSync(cachePath, 'utf-8');
    return JSON.parse(cache);
  }
  return [];
}

export function getPhotosFromCache(): Post[] {
  const cachePath = path.join(process.cwd(), 'photos-cache.json');
  if (fs.existsSync(cachePath)) {
    const cache = fs.readFileSync(cachePath, 'utf-8');
    return JSON.parse(cache);
  }
  return [];
}

export async function fetchPublishedPosts(databaseID: string) {
  // This function is now intended to be used only by the caching script.
  const posts = await notion.databases.query({
    database_id: databaseID!,
    filter: {
      and: [
        {
          property: 'Status',
          status: {
            equals: 'Published',
          },
        },
      ],
    },
    sorts: [
      {
        property: 'Published Date',
        direction: 'descending',
      },
    ],
  });

  return posts.results as PageObjectResponse[];
}

export async function getPost(slug: string): Promise<Post | null> {
  const posts = getPostsFromCache();
  const post = posts.find((p) => p.slug === slug);
  return post || null;
}

export async function getPostFromNotion(pageId: string): Promise<Post | null> {
  try {
    const page = (await notion.pages.retrieve({
      page_id: pageId,
    })) as PageObjectResponse;
    const mdBlocks = await n2m.pageToMarkdown(pageId);

    const { parent: contentString } = n2m.toMarkdownString(mdBlocks);

    const properties = page.properties;

    const imageName = properties['Image Name'];
    const title = properties['Title'];
    const slug = properties['Slug'];
    const excerpt = properties['Excerpt'];
    const date = properties['Published Date'];

    const titleText =
      title.type === 'title' && title.title[0]?.plain_text ? title.title[0].plain_text : undefined;

    if (!titleText) {
      throw new Error(`Add a title for ${pageId}`);
    }

    const slugText =
      slug.type === 'rich_text' && slug.rich_text[0]?.plain_text
        ? slug.rich_text[0]?.plain_text
        : 'no-slug';

    if (!slugText) {
      throw new Error(`Add a slug for ${titleText}`);
    }

    const coverImage =
      imageName?.type === 'rich_text' && imageName.rich_text && imageName.rich_text[0]
        ? imageName.rich_text[0].plain_text
        : undefined;

    if (!coverImage) {
      throw new Error(`Add a cover image for ${titleText}`);
    }

    const excerptText =
      excerpt.type === 'rich_text' && excerpt.rich_text[0]?.plain_text
        ? excerpt.rich_text[0]?.plain_text
        : undefined;

    if (!excerptText) {
      throw new Error(`Add excerpt text for ${titleText}`);
    }

    const dateText = date.type === 'date' && date.date?.start ? date.date?.start : undefined;

    if (!dateText) {
      throw new Error(`Add a date for ${titleText}`);
    }

    const post: Post = {
      id: page.id,
      title: titleText,
      slug: slugText,
      excerpt: excerptText,
      coverImage,
      date: dateText,
      content: contentString,
    };

    return post;
  } catch (error) {
    console.error('Error getting post:', error);
    return null;
  }
}
