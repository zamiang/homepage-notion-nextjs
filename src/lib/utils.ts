import { type ClassValue, clsx } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function calculateReadingTime(wordCount: number): string {
  const WORDS_PER_MINUTE = 225; // Average adult reading speed
  const minutes = Math.ceil(wordCount / WORDS_PER_MINUTE);
  return `${minutes} min read`;
}
