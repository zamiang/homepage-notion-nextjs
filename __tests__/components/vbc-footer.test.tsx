import VBCFooter from '@/components/vbc-footer';
import * as notionModule from '@/lib/notion';
import { Post } from '@/lib/notion';
import { render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

// Mock the notion module
vi.mock('@/lib/notion', async () => {
  const actual = await vi.importActual<typeof notionModule>('@/lib/notion');
  return {
    ...actual,
    getVBCSectionPostsPostsFromCache: vi.fn(),
  };
});

const mockVBCPosts: Post[] = [
  {
    id: '1',
    title: 'A First Post',
    slug: 'first-post',
    coverImage: 'cover1.jpg',
    date: '2023-01-01',
    excerpt: 'First VBC post.',
    content: 'Content 1',
    author: 'Test Author',
    section: 'VBC',
  },
  {
    id: '2',
    title: 'B Second Post',
    slug: 'second-post',
    coverImage: 'cover2.jpg',
    date: '2023-01-02',
    excerpt: 'Second VBC post.',
    content: 'Content 2',
    author: 'Test Author',
    section: 'VBC',
  },
  {
    id: '3',
    title: 'C Third Post',
    slug: 'third-post',
    coverImage: 'cover3.jpg',
    date: '2023-01-03',
    excerpt: 'Third VBC post.',
    content: 'Content 3',
    author: 'Test Author',
    section: 'VBC',
  },
];

describe('VBCFooter', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    vi.mocked(notionModule.getVBCSectionPostsPostsFromCache).mockReturnValue(mockVBCPosts);
  });

  it('should render VBC series title', () => {
    render(<VBCFooter slug="second-post" />);
    expect(screen.getByText('Why Value-Based Care is Harder Than Rocket Science')).toBeInTheDocument();
  });

  it('should render VBC series description', () => {
    render(<VBCFooter slug="second-post" />);
    expect(
      screen.getByText(/This series argues that U\.S\. healthcare is "harder than rocket science"/),
    ).toBeInTheDocument();
  });

  it('should render all VBC posts', () => {
    render(<VBCFooter slug="second-post" />);
    expect(screen.getByText('A First Post')).toBeInTheDocument();
    expect(screen.getByText('B Second Post')).toBeInTheDocument();
    expect(screen.getByText('C Third Post')).toBeInTheDocument();
  });

  it('should sort posts alphabetically by title', () => {
    render(<VBCFooter slug="second-post" />);
    const postTitles = screen
      .getAllByRole('heading', { level: 4 })
      .map((h) => h.textContent)
      .filter((t) => t && !t.includes('wide business')); // Exclude future post

    expect(postTitles).toEqual(['A First Post', 'B Second Post', 'C Third Post']);
  });

  it('should mark correct post as current', () => {
    const { container } = render(<VBCFooter slug="second-post" />);
    const currentPost = container.querySelector('.current-post');
    expect(currentPost).toBeInTheDocument();
    expect(currentPost?.textContent).toContain('B Second Post');
  });

  it('should mark posts before current as past', () => {
    const { container } = render(<VBCFooter slug="second-post" />);
    const pastPosts = container.querySelectorAll('.past-post');
    expect(pastPosts).toHaveLength(1);
    expect(pastPosts[0]?.textContent).toContain('A First Post');
  });

  it('should mark post after current as next', () => {
    const { container } = render(<VBCFooter slug="second-post" />);
    const nextPost = container.querySelector('.next-post');
    expect(nextPost).toBeInTheDocument();
    expect(nextPost?.textContent).toContain('C Third Post');
  });

  it('should render future post placeholder', () => {
    render(<VBCFooter slug="second-post" />);
    expect(screen.getByText('September 2025')).toBeInTheDocument();
    expect(
      screen.getByText('The wide business: VBC through the lens of operations research'),
    ).toBeInTheDocument();
  });

  it('should render future post with correct class', () => {
    const { container } = render(<VBCFooter slug="second-post" />);
    const futurePosts = container.querySelectorAll('.future-post');
    expect(futurePosts).toHaveLength(1);
  });

  it('should handle first post in series', () => {
    const { container } = render(<VBCFooter slug="first-post" />);
    const currentPost = container.querySelector('.current-post');
    expect(currentPost?.textContent).toContain('A First Post');

    const pastPosts = container.querySelectorAll('.past-post');
    expect(pastPosts).toHaveLength(0);

    const nextPost = container.querySelector('.next-post');
    expect(nextPost?.textContent).toContain('B Second Post');
  });

  it('should handle last post in series', () => {
    const { container } = render(<VBCFooter slug="third-post" />);
    const currentPost = container.querySelector('.current-post');
    expect(currentPost?.textContent).toContain('C Third Post');

    const pastPosts = container.querySelectorAll('.past-post');
    expect(pastPosts).toHaveLength(2);

    const nextPost = container.querySelector('.next-post');
    expect(nextPost).toBeNull();
  });

  it('should handle empty VBC posts', () => {
    vi.mocked(notionModule.getVBCSectionPostsPostsFromCache).mockReturnValue([]);

    render(<VBCFooter slug="any-slug" />);
    expect(screen.getByText('Why Value-Based Care is Harder Than Rocket Science')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { container } = render(<VBCFooter slug="second-post" />);
    expect(container).toMatchSnapshot();
  });
});
