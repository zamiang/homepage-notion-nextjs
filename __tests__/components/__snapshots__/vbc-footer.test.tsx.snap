// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VBCFooter > should match snapshot 1`] = `
<div>
  <div>
    <div
      class="center-divider"
    />
    <div
      class=""
    >
      <h3
        class="heading"
      >
        Why Value-Based Care is Harder Than Rocket Science
      </h3>
      <p>
        This series argues that U.S. healthcare is "harder than rocket science" due to its "consolidated fragmentation," where powerful, siloed players hinder effective, affordable patient care. The main conclusion is that in an era of AI and consolidation, we need a major shift in data policy to deliver on the promise of improved quality of care at reduced cost.
      </p>
      <div
        class="divider"
      />
      <div
        class="post past-post"
      >
        <h4
          style="margin-top: 0px;"
        >
          <a
            aria-label="A First Post"
            href="/writing/first-post"
          >
            A First Post
          </a>
        </h4>
      </div>
      <div
        class="post current-post"
      >
        <h4
          style="margin-top: 0px;"
        >
          <a
            aria-label="B Second Post"
            href="/writing/second-post"
          >
            B Second Post
          </a>
        </h4>
      </div>
      <div
        class="post next-post"
      >
        <a
          aria-label="Next post: C Third Post"
          class="next-button"
          href="/writing/third-post"
        >
          Next
        </a>
        <h4
          style="margin-top: 0px;"
        >
          <a
            aria-label="C Third Post"
            href="/writing/third-post"
          >
            C Third Post
          </a>
        </h4>
        <p
          class="truncated-text"
        >
          Third VBC post.
        </p>
      </div>
      <div
        class="post future-post"
      >
        <div
          class="next-button"
        >
          September 2025
        </div>
        <h4
          style="margin-top: 0px;"
        >
          The wide business: VBC through the lens of operations research
        </h4>
      </div>
    </div>
  </div>
</div>
`;
